server/matchmaker/algorithm.py

```diff
@@ -179,44 +170,105 @@ def _remove_duplicates(self) -> List[Match]:
         """ A graph in adjacency list representation, whose nodes are the searches
-        and whose edges are the top few possible matchings for each node.
+        and whose edges are the possible matchings for each node. Checks every
+        possible edge for inclusion in the graph.
 
         Note that the highest quality searches come at the end of the list so that
         it can be used as a stack with .pop().
+
+        Time complexity: O(n^2)
         """
-        return {
-            search: sorted(
-                _MatchingGraph._get_top_matches(
-                    search, filter(lambda s: s is not search, searches)
-                ),
-                key=lambda other: search.quality_with(other)
-            )
-            for search in searches
-        }
+        adj_list = {search: [] for search in searches}
+
+        # Generate every edge. There are 'len(searches) choose 2' of these.
+        for search, other in subset_pairs(searches):
+            quality = search.quality_with(other)
+            if not _MatchingGraph.is_possible_match(search, other, quality):
+                continue
+
+            # Add the edge in both directions
+            adj_list[search].append((other, quality))
+            adj_list[other].append((search, quality))
+
+        # Sort edges by their weights i.e. match quality
+        for search, neighbors in adj_list.items():
+            neighbors.sort(key=lambda edge: edge[1])
+
+        return adj_list
 
     @staticmethod
-    def _get_top_matches(search: Search, others: Iterable[Search]) -> List[Search]:
-        def is_possible_match(other: Search) -> bool:
-            quality_log_string = (
-                f"Quality between {search} and {other}: {search.quality_with(other)}"
-                f" thresholds: [{search.match_threshold}, {other.match_threshold}]."
-            )
+    def build_fast(searches: List[Search]) -> WeightedGraph:
+        """ Builds approximately the same graph as `build_full`, but does not
+        check every possible edge.
 
-            if search.matches_with(other):
-                _MatchingGraph._logger.debug(
-                    f"{quality_log_string} Will be considered during stable marriage."
-                )
-                return True
-            else:
-                _MatchingGraph._logger.debug(
-                    f"{quality_log_string} Will be discarded for stable marriage."
-                )
-                return False
+        Time complexity: O(n*log(n))
+        """
+        adj_list = {search: [] for search in searches}
+        # TODO: Support searches with larger parties
+        # Sort all searches by their average rating
+        searches = sorted(
+            searches, key=lambda s: s.ratings[0][0] - 3 * s.ratings[0][1]
+        )
+        # Now compute quality with `num_to_check` nearby searches on either side
+        num_to_check = max((4, int(math.log(len(searches), 2)))) // 2
```

Comment:
You chose that so that it checks the highest number of edges while still staying `O(n log n)`? Nice :)

Reply:
Yea I figured we might as well maximize the number of opponents checked while still keeping the time complexity the same.